cmd/spire-server/cli/healthcheck/healthcheck.go

```diff
@@ -17,82 +16,52 @@ func NewHealthCheckCommand() cli.Command {
 }
 
-func newHealthCheckCommand(env *common_cli.Env) *healthCheckCommand {
-	return &healthCheckCommand{
-		env:     env,
-		timeout: common_cli.DurationFlag(time.Second * 5),
-	}
+func newHealthCheckCommand(env *common_cli.Env) cli.Command {
+	return util.AdaptCommand(env, new(healthCheckCommand))
 }
 
 type healthCheckCommand struct {
-	env *common_cli.Env
-
-	socketPath string
-	timeout    common_cli.DurationFlag
-	shallow    bool
-	verbose    bool
+	shallow bool
+	verbose bool
 }
 
-func (c *healthCheckCommand) Help() string {
-	// ignoring parsing errors since "-h" is always supported by the flags package
-	_ = c.parseFlags([]string{"-h"})
-	return ""
+func (c *healthCheckCommand) Name() string {
+	return "healthcheck"
 }
 
 func (c *healthCheckCommand) Synopsis() string {
 	return "Determines server health status"
 }
 
-func (c *healthCheckCommand) Run(args []string) int {
-	if err := c.parseFlags(args); err != nil {
-		return 1
-	}
-	if err := c.run(); err != nil {
-		// Ignore error since a failure to write to stderr cannot very well be
-		// reported
-		_ = c.env.ErrPrintf("Server is unhealthy: %v\n", err)
-		return 1
-	}
-	if err := c.env.Println("Server is healthy."); err != nil {
-		return 1
-	}
-	return 0
-}
-
-func (c *healthCheckCommand) parseFlags(args []string) error {
-	fs := flag.NewFlagSet("health", flag.ContinueOnError)
-	fs.SetOutput(c.env.Stderr)
-	fs.StringVar(&c.socketPath, "registrationUDSPath", util.DefaultSocketPath, "Registration API UDS path")
+func (c *healthCheckCommand) AppendFlags(fs *flag.FlagSet) {
 	fs.BoolVar(&c.shallow, "shallow", false, "Perform a less stringent health check")
 	fs.BoolVar(&c.verbose, "verbose", false, "Print verbose information")
-	return fs.Parse(args)
 }
 
-func (c *healthCheckCommand) run() error {
-	if c.verbose {
-		if err := c.env.Println("Checking server health..."); err != nil {
-			return err
-		}
+func (c *healthCheckCommand) Run(ctx context.Context, env *common_cli.Env, client util.ServerClient) error {
+	if err := c.run(ctx, env, client); err != nil {
+		return fmt.Errorf("server is unhealthy: %w", err)
 	}
+	if err := env.Println("Server is healthy."); err != nil {
+		return err
+	}
+	return nil
+}
 
-	client, err := util.NewServerClient(c.socketPath)
-	if err != nil {
-		if c.verbose {
-			// Ignore error since a failure to write to stderr cannot very well
-			// be reported
-			_ = c.env.ErrPrintf("Failed to create client: %v\n", err)
+func (c *healthCheckCommand) run(ctx context.Context, env *common_cli.Env, client util.ServerClient) error {
+	if c.verbose {
+		if err := env.Println("Checking server health..."); err != nil {
+			return err
 		}
-		return errors.New("cannot create health client")
 	}
-	defer client.Release()
 
 	healthClient := client.NewHealthClient()
-	resp, err := healthClient.Check(context.Background(), &grpc_health_v1.HealthCheckRequest{})
+	resp, err := healthClient.Check(ctx, &grpc_health_v1.HealthCheckRequest{})
```

Comment:
A side effect of moving to the common command structure is that this context can now time out.... That is strictly a change in behavior from the healthcheck previously, which specified, but did not honor, the timeout flag.

Happy to hear any concerns about this change...

Reply:
Hah... the old fix-a-bug-create-a-bug conundrum :)

This should probably be called out individually in the CHANGELOG

Reply:
Right now health checks happen inline with the RPC call, which means if something hangs up, the timeout might help. That being said I could imagine we move to a interval-based check and the RPC call just grabbing the latest status, in which case a timeout would be unlikely.

One option is to purposefully ignore the passed in context with a big fat note why :man_shrugging:

Reply:
No I think it's good to respect it. Just was commenting that we should note the change in behavior.

> That being said I could imagine we move to a interval-based check and the RPC call just grabbing the latest status, in which case a timeout would be unlikely.

Yes I think we will ultimately need to do this for a number of reasons. IMO the design resulting from https://github.com/spiffe/spire/issues/2047 should work this way.